pna.p4

```diff
@@ -665,19 +665,75 @@ extern void mirror_packet(MirrorSlotId_t mirror_slot_id,
 // probably be a parameter that specifies the new entry's initial
 // expire_time_profile_id.
 
-extern bool add_entry<T>(string action_name,
-                         in T action_params,
-                         in ExpireTimeProfileId_t expire_time_profile_id);
-
-// If we support the following variant of add_entry, with no
-// expire_time_profile_id parameter, then we need to define its
-// behavior in terms of what the added entry's value of
-// expire_time_profile_id is.  One option would be to define the
-// behavior to be the same as the 3-parameter version of add_entry
-// above, with expire_time_profile_id equal to 0.
-
-//extern bool add_entry<T>(string action_name,
-//                         in T action_params);
+// The bit width of this type is allowed to be different for different
+// target devices.  It must be at least a 1-bit wide type.
+
+typedef bit<1> AddEntryErrorStatus_t;
+
+const AddEntryErrorStatus_t ADD_ENTRY_SUCCESS = 0;
+const AddEntryErrorStatus_t ADD_ENTRY_NOT_DONE = 1;
+
+// Targets may define target-specific non-0 constants of type
+// AddEntryErrorStatus_t if they wish.
+
+// add_entry() must be called within the body of a P4 action that is
+// the default_action of a table with the table property `add_on_miss`
+// equal to true.
+//
+// The key of the entry added will always be the same as the key that
+// was just looked up in the table, and experienced a miss.
+//
+// `action_name` is the name of an action of this table.  It must be
+// possible for this action to be the action of an entry added to the
+// table, i.e. it is an error if the action has the annotation
+// `@defaultonly`.
+//
+// Type T must be a struct type whose field names have the same name
+// as the parameters of the action being added, in the same order, and
+// have the same type as the corresponding action parameters.
+//
+// `action_params` will become the action parameters of the new entry
+// to be added.
+//
+// `expire_time_profile_id` is the initial expire time profile id of
+// the entry added.
+//
+// The return value will be ADD_ENTRY_SUCCESS if the entry was
+// successfully added, otherwise it will be some other value not equal
+// to ADD_ENTRY_SUCCESS.  Targets are allowed to define only one
+// failure return value, or several if they wish to provide more
+// detail on the reason for the failure to add the entry.
+
+extern AddEntryErrorStatus_t add_entry<T>(
+    string action_name,
+    in T action_params,
```

Comment:
Is this single action parameter ? Can we provide a list of action parameters here ? if not, then we should call action_param ?

Reply:
The type `T` is intended to be a struct containing as many fields as the corresponding action being added has action parameters, and for those struct fields to match one-to-one with the names and types of the action parameters.

There are likely other ways this could be defined in P4, but I do not know another good way.  It is not straightforward to define the signature of a function or method that takes any number of parameters today in P4, e.g. unlike Python, and probably C++.

Reply:
@jfingerh I get that. Hence the suggestion/wording to be  "action_param" versus "action_params"

Reply:
What if the action has 3 parameters?  You would not object if it was called `action_param` in that case, even though that one parameter to `add_entry` represents 3 parameters to the action?